### ISSUE TITLE
[Question] ConfigAbstractFactory: why is service id required to be a valid class

### ISSUE THREAD

**Issue by codeliner** [zendframework/zend-servicemanager#171]
Sorry for the following c&p text. Not much time atm but I thought it would be good to ask this question because we discussed the limitation of ConfigAbstractFactory together with @basz in a chat:

me: just a hint regarding Zend's ConfigAbstractFactory: It has a limitation because the factory assumes/requires that the service id is a valid class. I often use an interface as service id and let a factory return an implementation of that interface. In this case the ConfigAbstractFactory won't work or at least not without a workaround like setting the interface as alias for the class.
In Symfony you do:
```
'my_service_id':
  class: 'Acme\MyClass'
  arguments: ['dep1_service_id', 'dep2_service_id']
```
Same would work with the ConfigAbstractFactory, not sure why the shortcut is used: service_id = valid class

**Comment by mtymek** [comment]
I always create map factory to actual class name and then use alias for interface:

```
return [
            'service_manager' => [
                'factories' => [
                    FooService::class => SomeGenericFactory::class,
                ],
                'aliases' => [
                    FooServiceInterface::class => FooService::class,
                ],
            ],
```

Using this method allows me to easily switch between implementations just by editing alias. For example, I have an integration test that (among other things) sends an e-mail. Before test is run I can simply reconfigure SM to switch from SMTP transport to something that saves e-mail HTML in a variable.

**Comment by codeliner** [comment]
@mtymek I consider this a workaround

> workaround like setting the interface as alias for the class.

I switch the factory in this case ;)
Anyway, I have a good reason why I don't want to expose implementation details:

When I have a `UserRepository` interface, the model is not interested if it is a `MongoUserRepository` or a `PdoUserRepository`, hence I don't want to expose the concrete implementation as a service because my model should not have a direct access to it. I know that the service manager can not really help here but registering the concrete implementation as a service would hint another developer that they can use this implementation in let's say a domain service but that would be wrong.

To summarize: I'm not sure if I like the new `ConfigAbstractFactory`. The service manager works great without it: Service <-> Factory that's it!
Now you have to think about rules again: when should I use the ConfigAbstractFactory, when is it not possible due to limitiations, when should I not use it because a factory is clearly the better choice ....
Without the limitation you could decide to use `ConfigAbstractFactory` for all your services or always use factories or if you want to confuse others mix both concepts ;) At the moment the first option is not possible IMO.

**Comment by Ocramius** [comment]
@codeliner the `ConfigAbstractFactory` is entirely opt-in, so it's generally OK to just ignore it and use it only for concrete classes.

**Comment by codeliner** [comment]
@Ocramius 

> so it's generally OK to just ignore it and use it only for concrete classes.

sure, but why this limitation? Modified example from the docs:

```php
use Zend\ServiceManager\AbstractFactory\ConfigAbstractFactory;

return [
    ConfigAbstractFactory::class => [
        MyInvokableClass::class => [],
        MySimpleClass::class => [
            PSR\LoggerInterface::class,   <--- only dep is a PSR-3 LoggerInterface, not interested in more than that       
        ],
        PSR\LoggerInterface::class => [
           'class' => MyPsrLogger::class, <--- can be omitted, if service id = class
           'arguments' =>  Handler::class,
        ],
    ],
];
```

**Comment by codeliner** [comment]
Additional background:
@prooph we make heavy use of factories + array-config to be able to provide a framework agnostic way of setting up our components. But people struggling with it. The funny thing is, it is always called Zend's Array-Config, not PHP's Array-Config ;)
Anyway, people are confused by examples using `MyClass::class` as service id in the interop-container configuration but as "use this class" in a service configuration because it looks the same but has another purpose.
Same happend to me with the `ConfigAbstractFactory`. I started to port some deps-only services to use this new factory and removed my own factories. It worked for two services and the third one used an Interface as service id, so :boom: :cry:

**Comment by weierophinney** [comment]
The configuration was designed for a very specific, very narrow use case, deliberately, in order to make the configuration as flat as possible. @GeeH hopefully can jump in to verify/validate what I describe below, as this feature was his contribution.

The configuration you recommend above, @codeliner, is similar to configuration we have used in the past for plugin loaders and zend-di, and requires:

- "magic" keys for detailing which items are the class to instantiate, and which are the arguments/options to pass when instantiating.
- additional nesting of the arguments; they get pushed another level deeper.

We specifically did not want either of these. We could have omitted the keys by using class constants, but that leads to either additional class imports in your configuration so you can have short key names (e.g. `ConfigAbstractFactory::KEY_CLASS`), fully qualified class name + constant name notation (e.g. `Zend\ServiceManager\AbstractFactory\ConfigAbstractFactory::KEY_CLASS`), or falling back to the key values themselves (potentially risky should we change them). Having an additional level of nesting is undesirable, as it becomes harder to associate what is being instantiated with the arguments for instantiation.

The notation chosen:

```php
ClassName::class => [
    FirstArgument::class,
    SecondArgument::class
]
```

retains some affinity to how you might instantiate the class:

```php
new ClassName(
    new FirstArgument(),
    new SecondArgument()
)
```

We have been recommending using FQCN for service names as that makes it clear the _type_ of instance being created. It may not actually be an instance of _exactly_ that type, but chances are that you are creating and returning something that _fulfills_ that type. Our observations have actually been contrary to what you report; most developers have indicated that this is easier to understand than arbitrary string names.

Which leads to the argument for using interfaces as key names. Clearly, the way that the `ConfigAbstractFactory` consumes configuration will not support that (nor will it for the `ReflectionBasedAbstractFactory`) as the _key_ name is the _class_ to instantiate. However, that's precisely what _aliasing_ is for in zend-servicemanager. You would create one or more `ConfigAbstractFactory` entries for the various _concrete_ service types you support, and then _alias_ the interface to the appropriate concrete service type.

If you prefer mapping the interface service name to a factory, then generate a factory using `vendor/bin/generate-factory-for-class`, which will generate a factory using reflection; you can then map that factory to your interface name. That tool exists both to improve performance (no configuration needs to be read or reflection done in a production environment) as well as to support use cases such as what you've demonstrated above.

**Comment by GeeH** [comment]
I can't add anything as @weierophinney has covered it all, but to reiterate, the `ConfigAbstractFactory` has a very deliberate use case, and that use case is where you have a very simple class with basic dependencies. It was build to save you creating and maintaining hundreds of classes which were essentially: 

```
namespace My\Service\Factory;

class MyServiceFactory
{
    public function __invoke($serviceManager)
    {
        return new MyService(
            $serviceManager->get(AnotherClass::class),
            $serviceManager->get(YetAnotherClass::class),
        );
    }
}
```

That's the use case (on purpose). If you need more complexity either you are encouraged to create you own actual factories (maybe using the new command line tool?) or to create you own abstract factory.

**Comment by Ocramius** [comment]
@GeeH you closed it, you own it.

**Comment by GeeH** [comment]
@Ocramius  If only you could code as efficiently as you can administrate.

**Comment by Ocramius** [comment]
I do, are you envious ;-)

**Comment by codeliner** [comment]
thank you all for the detailed response :+1: I'm going to build my own factory now based on the idea of the `ConfigAbstractFactory` but with support for:

```php
namespace My\Service\Factory;

class MyRepositoryFactory
{
    public function __invoke($serviceManager): MyRepositoryInterfaceUsedAsServiceId
    {
        return new MyConcreteRepository(
            $serviceManager->get(AnotherClass::class),
            $serviceManager->get(YetAnotherClass::class),
        );
    }
}
```

;)

@weierophinney 

> Our observations have actually been contrary to what you report; most developers have indicated that this is easier to understand than arbitrary string names.

The problem is more about yml/xml config versus PHP array-config ;) Users who are familiar with frameworks like Symfony are not used to use the language feature `::class` in configuration files and they struggle with it and the factory concept at-all. 
Anyway, I like the idea of the `ConfigAbstractFactory` but not its limitation.

I'd like to have these new rules in my codebase:

1. If a service (independent of the service id used in the container) only requires other dependencies, configure it in the `MyExtendedConfigAbstractFactory`
2. If a service requires additional application configuration, use a dedicated factory implementing [interop-config](https://github.com/sandrokeil/interop-config)
